### ISSUE TITLE
Implement a new generation flow to generate low level service skelotons. 

### ISSUE THREAD

**Issue by aneeshafedo** [ballerina-platform/openapi-tools#1388]
**Description:**

**Describe your problem(s)**

**Describe your solution(s)**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

**Comment by NipunaRanasinghe** [comment]
@aneeshafedo the proposed sample looks good except for the below cases

1. can we use the name `request` instead of `incomingRequest` for the request paramter?
2. here the description should change as `get pet details by id` or something similae
<img width="296" alt="Screenshot 2023-06-09 at 15 01 19" src="https://github.com/ballerina-platform/openapi-tools/assets/29032600/1d9ebd4c-f3db-49fd-88e2-022c325649c4">

**Comment by lnash94** [comment]
How about generating resource functions without a return type? This would allow users to have different return values without causing an `error` in the Ballerina type. 
If we generate without a return, the user can have more control to change their return values according to usage.
@aneeshafedo, @NipunaRanasinghe WDYT?

**Comment by aneeshafedo** [comment]
> @aneeshafedo the proposed sample looks good except for the below cases
> 
> 1.  can we use the name `request` instead of `incomingRequest` for the request paramter?
> 2.  here the description should change to `get pet details by id` or something similar IMO
> 
> ![Screenshot 2023-06-09 at 15 01 19](https://user-images.githubusercontent.com/29032600/244654917-1d9ebd4c-f3db-49fd-88e2-022c325649c4.png)

Updated. Thank you

**Comment by lnash94** [comment]
IMO `--low-level-service` flag is good for this scenario, as I feel that `--with-low-level-service` suggests that we will generate two service files—one including low-level services and the other being the service we would typically generate.

**Comment by aneeshafedo** [comment]
> How about generating resource functions without a return type? This would allow users to have different return values without causing an `error` in the Ballerina type. If we generate without a return, the user can have more control to change their return values according to usage. @aneeshafedo, @NipunaRanasinghe WDYT?
> 
> ```ballerina
> resource function put pet(http:Caller caller, http:Request incomingRequest)  {
>     }
> ```

IMO the return type `error?` is also suitable since it does not mean that the method should always return a value and it aligns with Ballerina method signature structure.

**Comment by lnash94** [comment]
> > How about generating resource functions without a return type? This would allow users to have different return values without causing an `error` in the Ballerina type. If we generate without a return, the user can have more control to change their return values according to usage. @aneeshafedo, @NipunaRanasinghe WDYT?
> > ```ballerina
> > resource function put pet(http:Caller caller, http:Request incomingRequest)  {
> >     }
> > ```
> 
> IMO the return type `error?` is also suitable since it does not mean that the method should always return a value and it aligns with Ballerina method signature structure.

Yeah, that is true 🙂,  I'm wondering regarding the status code that maps to the `error` type will affect the user API implementation later.  There can be scenarios errors can be overridden. reference https://discord.com/channels/957996897782616114/1108326862079406101/1108519895072768090

**Comment by hevayo** [comment]
An alternative flag would be `--without-data-binding` which might be more relatable to the user.

**Comment by NipunaRanasinghe** [comment]
> IMO the return type `error?` is also suitable since it does not mean that the method should always return a value and it aligns with Ballerina method signature structure.

yeah it won't do any harm except assuming that all the resources can return errors by default (IINM we model ballerina error as 500 internal errors when generating the contracts as well).
as we focus on a "minimal representation" for the service here, having no return types will also make sense IMO.

**Comment by NipunaRanasinghe** [comment]
> An alternative flag would be `--without-data-binding` which might be more relatable to the user.

yeah this flag looks more intuitive. +1

**Comment by aneeshafedo** [comment]
> > IMO the return type `error?` is also suitable since it does not mean that the method should always return a value and it aligns with Ballerina method signature structure.
> 
> yeah it won't do any harm except assuming that all the resources can return errors by default (IINM we model ballerina error as 500 internal errors when generating the contracts as well). as we focus on a "minimal representation" for the service here, having no return types will also make sense IMO.

After a discussin with @NipunaRanasinghe decided to generate resources with no return type to be compatible with OpenAPI operation given. 

The final design is as follows. 

\`\`\`bal  
import ballerina/http;

listener http:Listener ep0 = new (443, config = {host: "demo.saas-app.com"});

service /v2 on ep0 {  
   # Update an existing pet  
   #  
   # + caller - The caller actions for responding to client requests  
   # + request - Represents an HTTP request  
   # + return - Optionally return an error   
   resource function put pet(http:Caller caller, http:Request request) {  
   }  
   # Update an existing pet  
   #  
   # + caller - The caller actions for responding to client requests  
   # + request - Represents an HTTP request  
   # + return - Optionally return an error   
   resource function post pet(http:Caller caller, http:Request request) {  
   }  
   # Get pet by Id  
   #  
   # + caller - The caller actions for responding to client requests  
   # + request - Represents an HTTP request  
   # + return - Optionally return an error   
   resource function get pet/\[int petId\](http:Caller caller, http:Request request) {  
   }  
}

\`\`\`

The cli command to generate a service with no data binding is 

`bal openapi -i <openapi-file> --mode service --without-data-binding`  
`bal openapi -i <openapi-file> --without-data-binding`

**Comment by menakaj** [comment]
@aneeshafedo @NipunaRanasinghe shall we keep the return type as `error?` by default? IMO that would make the signature more generic.

In mediation scenarios, we explicitly add return error? too. If we can get this by default, it would be much easier for us too. 

`resource function post pet(http:Caller caller, http:Request incomingRequest) returns error? `

**Comment by NipunaRanasinghe** [comment]
> @aneeshafedo @NipunaRanasinghe shall we keep the return type as `error?` by default? IMO that would make the signature more generic.
> 
> In mediation scenarios, we explicitly add return error? too. If we can get this by default, it would be much easier for us too.
> 
> `resource function post pet(http:Caller caller, http:Request incomingRequest) returns error? `

@menakaj +1 from me

**Comment by aneeshafedo** [comment]
> @aneeshafedo @NipunaRanasinghe shall we keep the return type as `error?` by default? IMO that would make the signature more generic.
> 
> In mediation scenarios, we explicitly add return error? too. If we can get this by default, it would be much easier for us too.
> 
> `resource function post pet(http:Caller caller, http:Request incomingRequest) returns error?`

Updated the fix with `returns error?`

**Comment by NipunaRanasinghe** [comment]
Closing with https://github.com/ballerina-platform/ballerina-lang/pull/40707 and https://github.com/ballerina-platform/openapi-tools/pull/1395
